File: src/downloads_organizer/topic_naming.py
from __future__ import annotations

import hashlib
import re
from collections import Counter
from urllib.parse import unquote, urlparse

from .config import COURSE_PATTERN
from .models import IndexedFile
from .text_features import STOPWORDS, tokenize

NOISE = STOPWORDS | {
    "v1", "v2", "v3", "draft", "updated", "update", "new", "old", "assignment",
    "solution", "solutions", "tutorial", "lab", "material", "materials", "main", "readme",
    "md", "pdf", "docx", "pptx", "txt",
}


def topic_key(files: list[IndexedFile], course_code: str = "") -> str:
    if course_code:
        return f"course:{course_code.upper()}"
    identity = "\n".join(sorted(file.fingerprint for file in files))
    return "cluster:" + hashlib.sha256(identity.encode()).hexdigest()[:20]


def _clean_tokens(value: str) -> list[str]:
    value = COURSE_PATTERN.sub(" ", value)
    value = re.sub(r"(?i)\b(?:v(?:ersion)?\s*)?\d+(?:\.\d+)*\b", " ", value)
    return [token for token in tokenize(value) if token not in NOISE]


def _ngrams(tokens: list[str]) -> list[tuple[str, ...]]:
    result = []
    for size in range(2, min(4, len(tokens)) + 1):
        result.extend(tuple(tokens[index:index + size]) for index in range(len(tokens) - size + 1))
    return result


def _format_phrase(tokens: tuple[str, ...]) -> str:
    spellings = {"freertos": "FreeRTOS", "linux": "Linux", "ml": "ML", "mlb": "MLB"}
    words = [
        spellings.get(token, token.upper() if any(char.isdigit() for char in token) else token.title())
        for token in tokens
    ]
    return " ".join(words)[:80]


def display_name(files: list[IndexedFile], course_code: str = "") -> str:
    if course_code:
        return course_code.upper()

    sources = []
    document_tokens = []
    for file in files:
        stem_tokens = _clean_tokens(file.path.stem)
        title_tokens = _clean_tokens(file.title)
        sources.extend([stem_tokens, title_tokens])
        document_tokens.append(set(stem_tokens + title_tokens + file.keywords + tokenize(file.text[:20_000])))

    stem_lists = [_clean_tokens(file.path.stem) for file in files]
    stem_common: tuple[str, ...] = ()
    if stem_lists:
        common = set(stem_lists[0]).intersection(*(set(tokens) for tokens in stem_lists[1:]))
        stem_common = tuple(dict.fromkeys(token for token in stem_lists[0] if token in common))[:4]

    source_lists = [
        _clean_tokens(unquote(urlparse(url).path))
        for file in files for url in file.source_urls
    ]
    if len(stem_common) <= 1 and source_lists:
        common = set(source_lists[0]).intersection(*(set(tokens) for tokens in source_lists[1:]))
        source_common = tuple(dict.fromkeys(token for token in source_lists[0] if token in common))[:4]
        if len(source_common) >= 2:
            return _format_phrase(source_common)
    if stem_common:
        return _format_phrase(stem_common)

    title_lists = [_clean_tokens(file.title) for file in files]
    if title_lists:
        common = set(title_lists[0]).intersection(*(set(tokens) for tokens in title_lists[1:]))
        stable = tuple(
            token for token in title_lists[0]
            if token in common and (any(char.isdigit() for char in token) or token in {"freertos", "linux"})
        )[:4]
        if stable:
            return _format_phrase(stable)

    candidates = Counter(ngram for source in sources for ngram in _ngrams(source))
    ranked: list[tuple[float, tuple[str, ...]]] = []
    for phrase, origin_count in candidates.items():
        coverage = sum(set(phrase).issubset(tokens) for tokens in document_tokens)
        if coverage < max(2, len(files) // 2 + 1):
            continue
        length_bonus = {2: 2.0, 3: 1.2}.get(len(phrase), 0.3)
        ranked.append((coverage * 10 + origin_count * 2 + length_bonus, phrase))
    if ranked:
        _, phrase = max(ranked, key=lambda item: item[0])
        return _format_phrase(phrase)

    if title_lists:
        common = set(title_lists[0]).intersection(*(set(tokens) for tokens in title_lists[1:]))
        ordered = tuple(token for token in title_lists[0] if token in common)[:4]
        if ordered:
            return _format_phrase(ordered)

    informative_titles = [tokens for tokens in title_lists if tokens]
    if informative_titles:
        return _format_phrase(tuple(min(informative_titles, key=lambda tokens: (len(tokens), tokens))[:4]))
    return "Related Documents"

File: Sources/TopicTidyCore/Clustering/TopicNaming.swift
import CryptoKit
import Foundation

public enum TopicNaming {
    static let noise: Set<String> = TextFeatures.stopwords.union([
        "v1", "v2", "v3", "draft", "updated", "update", "new", "old", "assignment",
        "solution", "solutions", "tutorial", "lab", "material", "materials", "main", "readme",
        "md", "pdf", "docx", "pptx", "txt",
    ])

    private static let versionRegex = try! NSRegularExpression(
        pattern: "(?i)\\b(?:v(?:ersion)?\\s*)?\\d+(?:\\.\\d+)*\\b"
    )

    /// `topic_key` is durable identity; `display_name` is editable presentation.
    public static func topicKey(_ files: [IndexedFile], courseCode: String = "") -> String {
        if !courseCode.isEmpty { return "course:\(courseCode.uppercased())" }
        let identity = files.map(\.fingerprint).sorted().joined(separator: "\n")
        let digest = SHA256.hash(data: Data(identity.utf8))
        return "cluster:" + digest.map { String(format: "%02x", $0) }.joined().prefix(20)
    }

    static func cleanTokens(_ value: String) -> [String] {
        let range = NSRange(value.startIndex..<value.endIndex, in: value)
        let withoutCourses = CoursePattern.regex.stringByReplacingMatches(
            in: value, options: [], range: range, withTemplate: " "
        )
        let fullRange = NSRange(withoutCourses.startIndex..<withoutCourses.endIndex, in: withoutCourses)
        let withoutVersions = versionRegex.stringByReplacingMatches(
            in: withoutCourses, options: [], range: fullRange, withTemplate: " "
        )
        return TextFeatures.tokenize(withoutVersions).filter { !noise.contains($0) }
    }

    static func ngrams(_ tokens: [String]) -> [[String]] {
        var result: [[String]] = []
        let maximum = min(4, tokens.count)
        guard maximum >= 2 else { return result }
        for size in 2...maximum {
            for index in 0...(tokens.count - size) {
                result.append(Array(tokens[index..<(index + size)]))
            }
        }
        return result
    }

    static func formatPhrase(_ tokens: [String]) -> String {
        let spellings = ["freertos": "FreeRTOS", "linux": "Linux", "ml": "ML", "mlb": "MLB"]
        let words = tokens.map { token -> String in
            if let spelling = spellings[token] { return spelling }
            if token.contains(where: { $0.isNumber }) { return token.uppercased() }
            return titleCase(token)
        }
        return Py.prefix(words.joined(separator: " "), 80)
    }

    /// `str.title()` for the single tokens produced by `tokenize`.
    static func titleCase(_ token: String) -> String {
        var result = ""
        var previousWasCased = false
        for character in token {
            if character.isLetter {
                result += previousWasCased ? String(character).lowercased() : String(character).uppercased()
                previousWasCased = true
            } else {
                result.append(character)
                previousWasCased = false
            }
        }
        return result
    }

    /// Order-preserving intersection: tokens of `first` that also appear in all others.
    static func commonTokens(_ lists: [[String]]) -> [String] {
        guard let first = lists.first else { return [] }
        var common = Set(first)
        for list in lists.dropFirst() { common.formIntersection(Set(list)) }
        var seen: Set<String> = []
        var ordered: [String] = []
        for token in first where common.contains(token) && !seen.contains(token) {
            seen.insert(token)
            ordered.append(token)
        }
        return ordered
    }

    public static func displayName(_ files: [IndexedFile], courseCode: String = "") -> String {
        if !courseCode.isEmpty { return courseCode.uppercased() }

        var sources: [[String]] = []
        var documentTokens: [Set<String>] = []
        for file in files {
            let stemTokens = cleanTokens(file.path.deletingPathExtension().lastPathComponent)
            let titleTokens = cleanTokens(file.title)
            sources.append(stemTokens)
            sources.append(titleTokens)
            documentTokens.append(Set(stemTokens + titleTokens + file.keywords
                + TextFeatures.tokenize(Py.prefix(file.text, 20_000))))
        }

        let stemLists = files.map { cleanTokens($0.path.deletingPathExtension().lastPathComponent) }
        var stemCommon: [String] = []
        if !stemLists.isEmpty {
            stemCommon = Array(commonTokens(stemLists).prefix(4))
        }

        let sourceLists = files.flatMap { file in
            file.sourceURLs.map { cleanTokens(unquote(PyURL.parse($0).path)) }
        }
        if stemCommon.count <= 1 && !sourceLists.isEmpty {
            let sourceCommon = Array(commonTokens(sourceLists).prefix(4))
            if sourceCommon.count >= 2 { return formatPhrase(sourceCommon) }
        }
        if !stemCommon.isEmpty { return formatPhrase(stemCommon) }

        let titleLists = files.map { cleanTokens($0.title) }
        if !titleLists.isEmpty {
            let common = Set(commonTokens(titleLists))
            let stable = Array(titleLists[0].filter { token in
                common.contains(token) && (token.contains(where: { $0.isNumber })
                    || token == "freertos" || token == "linux")
            }.prefix(4))
            if !stable.isEmpty { return formatPhrase(stable) }
        }

        var candidates = OrderedCounter<[String]>()
        for source in sources {
            for phrase in ngrams(source) { candidates.add(phrase) }
        }
        var ranked: [(score: Double, phrase: [String])] = []
        for phrase in candidates.keys {
            let originCount = candidates.count(of: phrase)
            let phraseSet = Set(phrase)
            let coverage = documentTokens.reduce(0) { $0 + (phraseSet.isSubset(of: $1) ? 1 : 0) }
            if coverage < max(2, files.count / 2 + 1) { continue }
            let lengthBonus: Double = phrase.count == 2 ? 2.0 : (phrase.count == 3 ? 1.2 : 0.3)
            ranked.append((Double(coverage) * 10 + Double(originCount) * 2 + lengthBonus, phrase))
        }
        // `max(..., key=...)` keeps the first maximal element.
        var best: (score: Double, phrase: [String])?
        for item in ranked where best == nil || item.score > best!.score {
            best = item
        }
        if let best { return formatPhrase(best.phrase) }

        if !titleLists.isEmpty {
            let ordered = Array(commonTokens(titleLists).prefix(4))
            if !ordered.isEmpty { return formatPhrase(ordered) }
        }

        let informativeTitles = titleLists.filter { !$0.isEmpty }
        if let shortest = informativeTitles.min(by: { left, right in
            if left.count != right.count { return left.count < right.count }
            return compareTokenLists(left, right) < 0
        }) {
            return formatPhrase(Array(shortest.prefix(4)))
        }
        return "Related Documents"
    }

    /// Python list comparison: element-wise, then by length.
    static func compareTokenLists(_ left: [String], _ right: [String]) -> Int {
        for index in 0..<min(left.count, right.count) {
            let order = Py.compare(left[index], right[index])
            if order != 0 { return order }
        }
        if left.count == right.count { return 0 }
        return left.count < right.count ? -1 : 1
    }
}
